api/src/main/java/org/openmrs/module/pihcore/event/PatientUpdateEventConsumer.java

```diff
@@ -0,0 +1,209 @@
+            if (value != null) {
+                SqlBuilder sql = new SqlBuilder();
+                sql.select("p.patient_id").from("patient", "p");
+                if (keyTable.equals("person")) {
+                    sql.where("p.patient_id = ?");
+                }
+                else {
+                    sql.innerJoin(keyTable, "x", "patient_id", "p", "patient_id");
+                    sql.where("x.").append(keyTable).append(" = ").append("?");
+                }
+                Integer patientId = database.executeQuery(sql.toString(), new ScalarHandler<>(1), value);
+                if (patientId == null) {
+                    if (!keyTable.equals("person")) {
+                        throw new RuntimeException("Unable to retrieve patient_id for: " + event);
+                    }
+                }
+                else {
+                    ret.add(patientId);
+                    if (!event.getTable().equals("relationship") || ret.size() == 2) {
+                        break;
+                    }
+                }
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * Once a particular event is connected ot a particular patient, this tracks this change for the patient
+     * @param patientId the patient id related to the event
+     * @param event the event to track
+     */
+    public void handlePatientEvent(Integer patientId, DbEvent event) {
+        if (log.isDebugEnabled()) {
+            log.debug("Patient Event: (patient_id=" + patientId + ") - " + event);
+        }
+        Timestamp lastUpdated = new Timestamp(event.getTimestamp());
+        String sql = "insert into dbevent_patient (patient_id, last_updated) values (?, ?) on duplicate key update last_updated = ?";
+        database.executeUpdate(sql, patientId, lastUpdated, lastUpdated);
+        if (event.getTable().equals("patient")) {
+            if (event.getOperation() == Operation.DELETE || event.getValues().getBoolean("voided")) {
+                sql = "update dbevent_patient set deleted = true where patient_id = ? and deleted = false";
+                database.executeUpdate(sql, patientId);
+            }
+            else if (!event.getValues().getBoolean("voided")) {
+                sql = "update dbevent_patient set deleted = false where patient_id = ? and deleted = true";
+                database.executeUpdate(sql, patientId);
+            }
+        }
+    }
+
+    /**
+     * Rather than stream all initial READ events one by one, it is much more efficient to set up the initial state
+     * by querying the database directly to record the most recent date updated and deleted status for all patients.
+     * This method performs this operation, and returns the max last_updated in the database
+     */
+    protected synchronized void performInitialSnapshot() {
+        if (!snapshotInitialized) {
+            log.warn("Performing initial snapshot into dbevent_patient from patient");
```

Comment:
In my initial implementation, I didn't do this.  I streamed the snapshot READ events in just like with the create/update/delete events, and if the event was a READ, I updated the table based on the date_created, date_changed, and date_voided columns.  But this did not perform well (took too long) on large existing DBs like HUM.  So instead, I wrote this method to perform a custom snapshot with our end-goal in mind, using DB optimizations to arrive at the same results.  On my system, this completes in 5-10 minutes on a HUM DB, which is perfectly acceptable.  We can discuss whether we even need any initialized values, depending on our downstream use case.  It seemed good to try to generate the best initial representation of when the most recent update for a patient likely took place, and then refine that based on streaming event data.

Reply:
Yeah, we need to think about performance (but agree that 5-10 minute seems perfectly reasonable)